pyopencl/array.py

```diff
@@ -59,35 +58,77 @@
 _COMMON_DTYPE_CACHE = {}
 
 
+class DoubleDowncastWarning(UserWarning):
+    pass
+
+
+_DOUBLE_DOWNCAST_WARNING = (
+        "The operation you requested would result in a double-precisision "
+        "quantity according to numpy semantics. Since your device does not "
+        "support double precision, a single-precision quantity is being returned.")
+
+
 def _get_common_dtype(obj1, obj2, queue):
     if queue is None:
         raise ValueError("PyOpenCL array has no queue; call .with_queue() to "
                 "add one in order to be able to perform operations")
 
-    dsupport = has_double_support(queue.device)
+    allow_double = has_double_support(queue.device)
     cache_key = None
-    o1_dtype = obj1.dtype
-    try:
-        cache_key = (o1_dtype, obj2.dtype, dsupport)
-        return _COMMON_DTYPE_CACHE[cache_key]
-    except KeyError:
-        pass
-    except AttributeError:
-        # obj2 doesn't have a dtype
-        try:
-            tobj2 = type(obj2)
-            cache_key = (o1_dtype, tobj2, dsupport)
+    o1_dtype = getattr(obj1, "dtype", type(obj1))
+    o2_dtype = getattr(obj2, "dtype", type(obj2))
+    o1_is_array = isinstance(obj1, Array)
+    o2_is_array = isinstance(obj2, Array)
+
+    if o1_dtype is int:
+        # must be an unsized scalar
+        o1_dtype = np.array(obj1).dtype
+
+        # Er, what? Well, consider this, which is true as of numpy 1.23.1:
+        #
+        # np.find_common_type([int16], [int64]) == int16
+        #
+        # apparently because of some hare-brained no-upcasts-because-of-scalars
+        # rule which even numpy itself ignores when determining result dtypes
+        # of an operation on integers. For non-integers OTOH, it seems to heed
+        # its own advice, but *shrug*. So we'll tell numpy that scalar integers
+        # "aren't scalars" to give their types "a bit more weight". *shudders*
+        o1_is_array = True
+
+    if o2_dtype is int:
+        # must be an unsized scalar
+        o2_dtype = np.array(obj2).dtype
+
+        # See above.
+        o2_is_array = True
+
+    cache_key = (o1_dtype, o2_dtype, o1_is_array, o2_is_array, allow_double)
+    #try:
+    #    return _COMMON_DTYPE_CACHE[cache_key]
+    #except KeyError:
+    #    pass
```

Comment:
@inducer Is this meant to be commented out?

Reply:
:facepalm: Thanks for pointing this out. I'll fix this, worry about the incompatibility in boxtree that broke things, and roll a new release.

Reply:
https://gitlab.tiker.net/inducer/pyopencl/-/jobs/457531/raw is the boxtree failure I mean, though now I'm no longer sure this is with current pyopencl? The pytential failure may be related...

Reply:
Ah! I got that in the array context branch and thought it was my fault for moving things around carelessly.

Yeah, the pytential issue can also be some dtype thing tripping up a tolerance check, so hopefully not too hard to fix :D

Reply:
If nothing else, `pytential` passes (locally) with this reverted..

Reply:
How extraordinarily sketchy. There shouldn't be a functional difference between use-the-cache and not. :facepalm:

Reply:
Ah, no, I meant with this whole changeset (went back to 78ee8f95), not just this commented out chunk.

Reply:
#609 should mop this up.

Reply:
https://github.com/inducer/pyopencl/releases/tag/v2022.2.3 is winding its way onto the package index.